test/NetTopologySuite.Samples.Console/Technique/DatelineUtility.cs

```diff
@@ -0,0 +1,233 @@
+        public static bool IsCrossesDateline(Geometry geometry)
+        {
+            System.Diagnostics.Debug.Print("geomType={0}", geometry.GeometryType);
+
+            // Argument checking
+            if (geometry == null)
+                throw new ArgumentNullException(nameof(geometry));
+
+            // If we don't have a geographic coordinate system 
+            if (!IsGeographicCS(geometry))
+                throw new ArgumentException("Geography does not have a geographic coordinate system.", nameof(geometry));
+
+            // Can't do anything with empty geometries
+            if (geometry.IsEmpty)
+                return false;
+
+            // If we have a geometry that is only consisting of points we don't need to do any
+            // dateline handling
+            if (geometry is IPuntal)
+                return false;
+
+            // Build filter that tests if we need to perform special dateline handling
+            var cdf = new IsCrossesDatelineFilter();
+            geometry.Apply(cdf);
+
+            // return the result
+            return cdf.IsCrossesDateline;
+        }
+
+        /// <summary>
+        /// Tests if the geometry has a geographic coordinate system
+        /// </summary>
+        /// <param name="geometry">The geometry</param>
+        /// <returns><c>true</c> if it has a geographic coordinate system, <c>false</c> otherwise.</returns>
+        private static bool IsGeographicCS(Geometry geometry)
+        {
+            // TODO: There are more geographic coordinate systems than WGS84. A filter of SRID.csv resulted in ~684
+            // geographic coordinate systems with a degree unit and Greenwich as the prime meridian. That is too many
+            // to list in this routine so the more popular SRIDs will be added as needed.
+            System.Diagnostics.Debug.Print("srid={0}", geometry.SRID);
+            return geometry.SRID == 4326;
+        }
+
+        /// <summary>
+        /// This is a filter that attempts to determine if the geometry crosses the dateline. Essentially, this is the key to the
+        /// entire dateline splitting utility.
+        /// </summary>
+        private class IsCrossesDatelineFilter : IGeometryFilter
+        {
+            public bool Done => IsCrossesDateline;
+
+            public bool GeometryChanged => false;
+
+            public bool IsCrossesDateline { get; private set; }
+
+            public void Filter(Geometry geometry)
+            {
+                // since we are filtering per geometry we could have multiple parts. so we must test each part individually for traversing
+                // dateline. if we were to test the CoordinateSequence with a IEntireCoordinateSequenceFilter filter then we could get false
+                // positives (ex: two islands on each side of the dateline).
+                System.Diagnostics.Debug.Print("num geoms={0}", geometry.NumGeometries);
```

Comment:
I'd argue that if a single `Polygon` of a `MultiPolygon` crosses the dateline you should consider the whole geometry to cross the dateline.